crates/wysiwyg/src/composer_model/format.rs

```diff
@@ -73,31 +73,77 @@ where
         }
     }
 
-    pub fn format(&mut self, format: InlineFormatType) -> ComposerUpdate<S> {
+    pub(crate) fn apply_pending_formats(&mut self, start: usize, end: usize) {
+        self.state
+            .toggled_format_types
+            .clone()
+            .iter()
+            .for_each(|format| {
+                if self.reversed_actions.contains(&format.action()) {
+                    self.format_range(start, end, format);
+                } else {
+                    self.unformat_range(start, end, format);
+                }
+            });
+        self.state.toggled_format_types.clear();
+    }
+
+    fn format(&mut self, format: InlineFormatType) -> ComposerUpdate<S> {
         // Store current Dom
         self.push_state_to_history();
         let (s, e) = self.safe_selection();
-        self.format_range(s, e, format);
-        self.create_update_replace_all()
+
+        if s == e {
+            self.toggle_zero_length_format(&format);
+            ComposerUpdate::update_menu_state(self.compute_menu_state())
+        } else {
+            self.format_range(s, e, &format);
+            self.create_update_replace_all()
+        }
     }
 
     fn format_range(
         &mut self,
         start: usize,
         end: usize,
-        format: InlineFormatType,
+        format: &InlineFormatType,
     ) {
+        assert!(start != end);
         let range = self.state.dom.find_range(start, end);
         self.format_several_nodes(&range, format);
     }
 
-    pub fn unformat(&mut self, format: InlineFormatType) -> ComposerUpdate<S> {
-        // Store current Dom
-        self.push_state_to_history();
+    fn unformat(&mut self, format: InlineFormatType) -> ComposerUpdate<S> {
         let (s, e) = self.safe_selection();
-        let range = self.state.dom.find_range(s, e);
-        self.unformat_several_nodes(s, e, &range, format);
-        self.create_update_replace_all()
+
+        if s == e {
+            self.toggle_zero_length_format(&format);
+            ComposerUpdate::update_menu_state(self.compute_menu_state())
+        } else {
+            self.unformat_range(s, e, &format);
+            self.create_update_replace_all()
+        }
+    }
+
+    fn unformat_range(
+        &mut self,
+        start: usize,
+        end: usize,
+        format: &InlineFormatType,
+    ) {
+        let range = self.state.dom.find_range(start, end);
+        self.unformat_several_nodes(start, end, &range, format);
+    }
+
+    fn toggle_zero_length_format(&mut self, format: &InlineFormatType) {
+        if self.state.toggled_format_types.contains(&format) {
+            self.state
+                .toggled_format_types
+                .iter()
+                .position(|f| f == format);
```

Comment:
@Hywan I hadn't seen this before - is this a normal way to search and modify an element in a Vec?

Reply:
Actually this is missing some code 🙈  

Should be something like that 

```rust
            let index = self.state
                .toggled_format_types
                .iter()
                .position(|f| f == format);
            if let Some(index) = index {
                self.state.toggled_format_types.remove(index);
            }
```

But perhaps there is a better alternative

Reply:
(which means we can remove the `if contains` in that case) 

/me is making test cases for multiple toggling cause it definitely doesn't work 😅